File: impl/go/cmd/liff/main_test.go
package main

import (
	"bytes"
	"strings"
	"testing"

	"liff"
)

func invoke(arguments ...string) (int, string, string) {
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	status := run(arguments, &stdout, &stderr)
	return status, stdout.String(), stderr.String()
}

func TestNoArgumentsPrintsRandomEntry(t *testing.T) {
	status, output, stderr := invoke()
	if status != 0 || stderr != "" {
		t.Fatalf("status = %d, stderr = %q", status, stderr)
	}
	word, definition, ok := strings.Cut(strings.TrimSuffix(output, "\n"), "\n")
	if !ok {
		t.Fatalf("random output has no definition: %q", output)
	}
	for _, entry := range liff.DefaultDictionary().Entries() {
		if entry.Word() == word {
			if entry.Definition() != definition {
				t.Errorf("definition differs for %s", word)
			}
			return
		}
	}
	t.Errorf("random word %q is not in the dictionary", word)
}

func TestFoundSearchesPrintEntry(t *testing.T) {
	tests := []struct {
		arguments []string
		prefix    string
	}{
		{[]string{"banteer"}, "BANTEER\nA lusty and raucous old ballad"},
		{[]string{"banteeer"}, "BANTEER\nA lusty and raucous old ballad"},
		{[]string{"glutt"}, "GLUTT LODGE\n"},
		{[]string{"bilb"}, "BILBSTER\n"},
		{[]string{"bil*"}, "BILBSTER\n"},
		{[]string{"b?lbster"}, "BILBSTER\n"},
		{[]string{"symonds", "yat"}, "SYMOND'S YAT\n"},
	}
	for _, testCase := range tests {
		status, output, stderr := invoke(testCase.arguments...)
		if status != 0 || stderr != "" || !strings.HasPrefix(output, testCase.prefix) {
			t.Errorf("run(%q) = status %d, stdout %q, stderr %q", testCase.arguments, status, output, stderr)
		}
	}
}

func TestAmbiguousAndLargeGlobOutput(t *testing.T) {
	status, output, _ := invoke("high")
	if status != 1 || output != "Did you mean?\nHIGH LIMERIGG\nHIGH OFFLEY\nAITH\nCHICAGO\n" {
		t.Errorf("ambiguous output = status %d, %q", status, output)
	}

	status, output, _ = invoke("b*")
	want := "Did you mean?\n" +
		"BABWORTH\nBALDOCK\nBALLYCUMBER\nBANFF\nBANTEER\n" +
		"BARSTIBLEY\nBAUGHURST\nBAUMBER\nBEALINGS\nBEAULIEU HILL\n" +
		"and 44 others\n"
	if status != 1 || output != want {
		t.Errorf("large glob output = status %d, %q", status, output)
	}

	status, output, _ = invoke("*")
	if status != 1 || !strings.HasSuffix(output, "and 540 others\n") || len(strings.Split(strings.TrimSuffix(output, "\n"), "\n")) != 12 {
		t.Errorf("all-entry glob output = status %d, %q", status, output)
	}
}

func TestExactlyElevenSuggestionsAreAllPrinted(t *testing.T) {
	status, output, _ := invoke("bo*")
	if status != 1 || strings.Contains(output, "and ") {
		t.Fatalf("eleven-result output = status %d, %q", status, output)
	}
	if lines := strings.Split(strings.TrimSuffix(output, "\n"), "\n"); len(lines) != 12 {
		t.Errorf("line count = %d, want 12", len(lines))
	}
}

func TestNotFoundHelpAndInvalidUsage(t *testing.T) {
	status, output, stderr := invoke("xyzzy")
	if status != 1 || output != "No definition found for \"xyzzy\".\n" || stderr != "" {
		t.Errorf("not found = status %d, stdout %q, stderr %q", status, output, stderr)
	}

	status, output, stderr = invoke("--help")
	if status != 0 || !strings.HasPrefix(output, "Usage: liff") || stderr != "" {
		t.Errorf("help = status %d, stdout %q, stderr %q", status, output, stderr)
	}

	status, output, stderr = invoke("--unknown")
	if status != 2 || output != "" || !strings.HasPrefix(stderr, "Usage: liff") {
		t.Errorf("invalid = status %d, stdout %q, stderr %q", status, output, stderr)
	}
}

File: impl/java/src/test/java/liff/cli/MainTest.java
package liff.cli;

import java.io.ByteArrayOutputStream;
import java.io.PrintWriter;
import java.nio.charset.StandardCharsets;
import liff.core.Liff;

/** Golden-output tests for the CLI adapter. */
public final class MainTest {
    private record Result(int status, String stdout, String stderr) {}

    private MainTest() {}

    public static void main(String[] arguments) {
        randomFoundAndNotFound();
        suggestionsAndGlobBoundaries();
        helpAndInvalidUsage();
        System.out.println("MainTest: OK");
    }

    private static Result invoke(String... arguments) {
        ByteArrayOutputStream stdoutBytes = new ByteArrayOutputStream();
        ByteArrayOutputStream stderrBytes = new ByteArrayOutputStream();
        PrintWriter stdout = new PrintWriter(stdoutBytes, false, StandardCharsets.UTF_8);
        PrintWriter stderr = new PrintWriter(stderrBytes, false, StandardCharsets.UTF_8);
        int status = Main.run(arguments, stdout, stderr, Liff.dictionary(), bound -> 0);
        stdout.flush();
        stderr.flush();
        return new Result(status, stdoutBytes.toString(StandardCharsets.UTF_8),
                stderrBytes.toString(StandardCharsets.UTF_8));
    }

    private static void randomFoundAndNotFound() {
        Result random = invoke();
        equal(0, random.status(), "random status");
        check(random.stdout().startsWith("AASLEAGH\n"), "random output");
        equal("", random.stderr(), "random stderr");

        for (String query : new String[] {"banteer", "banteeer", "glutt", "bilb", "bil*"}) {
            Result found = invoke(query);
            equal(0, found.status(), "found status for " + query);
            equal("", found.stderr(), "found stderr for " + query);
        }
        check(invoke("glutt").stdout().startsWith("GLUTT LODGE\n"), "glutt match");
        check(invoke("bilb").stdout().startsWith("BILBSTER\n"), "bilb match");
        check(invoke("bil*").stdout().startsWith("BILBSTER\n"), "unique glob match");
        check(invoke("symonds", "yat").stdout().startsWith("SYMOND'S YAT\n"),
                "joined query");

        Result missing = invoke("xyzzy");
        equal(1, missing.status(), "not-found status");
        equal("No definition found for \"xyzzy\".\n", missing.stdout(), "not-found output");
        equal("", missing.stderr(), "not-found stderr");
    }

    private static void suggestionsAndGlobBoundaries() {
        Result ambiguous = invoke("high");
        equal(1, ambiguous.status(), "ambiguous status");
        equal("Did you mean?\nHIGH LIMERIGG\nHIGH OFFLEY\nAITH\nCHICAGO\n",
                ambiguous.stdout(), "ambiguous output");

        Result many = invoke("b*");
        equal(1, many.status(), "large glob status");
        equal("Did you mean?\n"
                        + "BABWORTH\nBALDOCK\nBALLYCUMBER\nBANFF\nBANTEER\n"
                        + "BARSTIBLEY\nBAUGHURST\nBAUMBER\nBEALINGS\nBEAULIEU HILL\n"
                        + "and 44 others\n",
                many.stdout(), "large glob output");

        Result eleven = invoke("bo*");
        equal(1, eleven.status(), "eleven glob status");
        check(!eleven.stdout().contains("and "), "eleven results must not truncate");
        equal(12L, eleven.stdout().lines().count(), "eleven-result line count");

        Result all = invoke("*");
        equal(1, all.status(), "all glob status");
        check(all.stdout().endsWith("and 540 others\n"), "all glob truncation");
        equal(12L, all.stdout().lines().count(), "all-result line count");
    }

    private static void helpAndInvalidUsage() {
        Result help = invoke("--help");
        equal(0, help.status(), "help status");
        check(help.stdout().startsWith("Usage: liff"), "help stdout");
        equal("", help.stderr(), "help stderr");

        Result invalid = invoke("--unknown");
        equal(2, invalid.status(), "invalid status");
        equal("", invalid.stdout(), "invalid stdout");
        check(invalid.stderr().startsWith("Usage: liff"), "invalid stderr");
    }

    private static void check(boolean condition, String message) {
        if (!condition) {
            throw new AssertionError(message);
        }
    }

    private static void equal(Object expected, Object actual, String message) {
        if (!expected.equals(actual)) {
            throw new AssertionError(message + ": got " + actual + ", want " + expected);
        }
    }
}
